components/org.wso2.ei.analytics.elk/src/main/java/org/wso2/ei/analytics/elk/services/PublisherThread.java

```diff
@@ -0,0 +1,120 @@
+* in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing,
+* software distributed under the License is distributed on an
+* "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+* KIND, either express or implied. See the License for the
+* specific language governing permissions and limitations
+* under the License.
+*/
+
+package org.wso2.ei.analytics.elk.services;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.elasticsearch.client.transport.TransportClient;
+
+import org.wso2.ei.analytics.elk.publisher.ElasticStatisticsPublisher;
+
+import java.util.ArrayList;
+import java.util.Map;
+
+public class PublisherThread extends Thread {
+
+    private static final Log log = LogFactory.getLog(PublisherThread.class);
+
+    // To stop running
+    private volatile boolean shutdownRequested = false;
+
+    private TransportClient client;
+
+    @Override
+    public void run() {
+
+        // While not shutdown
+        while (!(shutdownRequested)) {
+
+            if (ElasticStatisticsPublisher.getAllMappingsQueue().isEmpty()) {
+
+                try {
+                    // Sleep for 1 second if the queue is empty
+                    Thread.sleep(1000);
+                } catch (InterruptedException e) {
+                    log.warn("Publisher Thread interrupted");
+                    Thread.currentThread().interrupt();
+                }
+
+            } else {
+
+                ObjectMapper objectMapper = new ObjectMapper();
+
+                ArrayList<String> jsonStringList = new ArrayList<String>();
+
+
+                if (client.connectedNodes().isEmpty()) {
+
+                    log.error("No available Elasticsearch nodes to connect. Waiting for nodes... ");
+
+                    try {
+                        // Sleep for 5 seconds if no nodes are available
+                        Thread.sleep(5000);
+                    } catch (InterruptedException e) {
+                        log.warn("Publisher Thread interrupted");
+                        Thread.currentThread().interrupt();
+                    }
+
+                } else {
+
+                    // While the map queue is not empty
+                    while (!(ElasticStatisticsPublisher.getAllMappingsQueue().isEmpty())) {
+
+                        // Dequeue Map from the queue
+                        Map<String, Object> map = ElasticStatisticsPublisher.getAllMappingsQueue().poll();
+
+                        try {
+                            jsonStringList.add(objectMapper.writeValueAsString(map));
+                        } catch (JsonProcessingException e) {
+                            log.error("Cannot convert to json");
```

Comment:
Shall we log the stacktrace also.

Reply:
Changed all error logs to log the exception also